### PR TITLE
Fix circular mapped type instantiations for arrays and tuples

### LINKED ISSUE #27881: Circular mapped tuple hangs compiler
I ran into this by accident while working on [this question](https://stackoverflow.com/q/52786547).

**TypeScript Version:**  master (54a5be1860c7f0de4e6196926d329fbbbf9c6081)

**Search Terms:** circular mapped tuple hang (plus the function names in the stack trace)

**Code**
```ts
export type Circular<T> = {[P in keyof T]: Circular<T>};
type tup = [number, number, number, number];

function foo(arg: Circular<tup>): tup {
  return arg;
}
```

**Expected behavior:** No hang.

**Actual behavior:** Compiler hangs with stack trace:
```
    at eval (eval at some (typescript/built/local/tsc.js:625:17), <anonymous>:1:1)
    at Object.some (typescript/built/local/tsc.js:625:17)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at typescript/built/local/tsc.js:41848:156
    at Object.some (typescript/built/local/tsc.js:627:25)
    at isTypeReferenceWithGenericArguments (typescript/built/local/tsc.js:41848:74)
    at getRelationKey (typescript/built/local/tsc.js:41886:17)
    at isTypeRelatedTo (typescript/built/local/tsc.js:40517:44)
    at checkTypeRelatedToAndOptionallyElaborate (typescript/built/local/tsc.js:39988:17)
    at checkTypeAssignableToAndOptionallyElaborate (typescript/built/local/tsc.js:39985:20)
    at checkReturnStatement (typescript/built/local/tsc.js:53735:25)
    at checkSourceElement (typescript/built/local/tsc.js:55220:28)
    at Object.forEach (typescript/built/local/tsc.js:182:30)
    at checkBlock (typescript/built/local/tsc.js:52831:20)
    at checkSourceElement (typescript/built/local/tsc.js:55199:28)
    at checkFunctionOrMethodDeclaration (typescript/built/local/tsc.js:52521:13)
    at checkFunctionDeclaration (typescript/built/local/tsc.js:52405:17)
    at checkSourceElement (typescript/built/local/tsc.js:55196:28)
    at Object.forEach (typescript/built/local/tsc.js:182:30)
    at checkSourceFileWorker (typescript/built/local/tsc.js:55400:20)
    at checkSourceFile (typescript/built/local/tsc.js:55372:13)
    at getDiagnosticsWorker (typescript/built/local/tsc.js:55446:17)
    at Object.getDiagnostics (typescript/built/local/tsc.js:55432:24)
    at typescript/built/local/tsc.js:85685:85
    at runWithCancellationToken (typescript/built/local/tsc.js:85651:24)
    at getSemanticDiagnosticsForFileNoCache (typescript/built/local/tsc.js:85674:20)
    at getAndCacheDiagnostics (typescript/built/local/tsc.js:85922:26)
    at getSemanticDiagnosticsForFile (typescript/built/local/tsc.js:85671:20)
    at typescript/built/local/tsc.js:85619:24
    at Object.flatMap (typescript/built/local/tsc.js:458:25)
    at getDiagnosticsHelper (typescript/built/local/tsc.js:85615:56)
    at Object.getSemanticDiagnostics (typescript/built/local/tsc.js:85626:20)
    at Object.emitFilesAndReportErrors (typescript/built/local/tsc.js:88873:46)
    at performCompilation (typescript/built/local/tsc.js:90800:29)
    at Object.executeCommandLine (typescript/built/local/tsc.js:90742:17)
    at Object.<anonymous> (typescript/built/local/tsc.js:90925:4)
    at Module._compile (module.js:650:14)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Function.Module.runMain (module.js:694:10)
    at startup (bootstrap_node.js:204:16)
    at bootstrap_node.js:625:3
```

**Playground Link:** N/A, the playground hangs before I can click the "Share" button. :smile: 

**Related Issues:** none found

### PR DESCRIPTION
**Pull request by ahejlsberg**
Mapped types can circularly reference themselves (because member resolution is deferred), but array and tuple type instantiations cannot. This causes an issue when a circular homomorphic mapped type is instantiated for an array or tuple type as we now create array and tuple instantiations for such mappings (see #26063). With is PR we quickly detect and stop the runaway recursion that can occur by substituting `errorType` for the instantiation. We would eventually do this after 50 levels of nested instantiations, but exponential recursive fan-out could keep us from ever getting there.

Fixes #27881.

### CODE REVIEW

File: tests/baselines/reference/objectLiteralNormalization.errors.txt [hunk 1]
```diff
@@ -52,9 +51,8 @@ tests/cases/conformance/expressions/objectLiterals/objectLiteralNormalization.ts
     a2 = { a: 1 };  // Error
     ~~
 !!! error TS2322: Type '{ a: number; }' is not assignable to type '{ a: number; b: number; } | { a: string; b?: undefined; } | { a?: undefined; b?: undefined; }'.
-!!! error TS2322:   Type '{ a: number; }' is not assignable to type '{ a?: undefined; b?: undefined; }'.
-!!! error TS2322:     Types of property 'a' are incompatible.
-!!! error TS2322:       Type 'number' is not assignable to type 'undefined'.
+!!! error TS2322:   Type '{ a: number; }' is not assignable to type '{ a: number; b: number; }'.
```
Comment (sandersn) [inline tests/baselines/reference/objectLiteralNormalization.errors.txt]: any idea why this changes? I don't see any mapped types here.

Comment (sandersn) [reply tests/baselines/reference/objectLiteralNormalization.errors.txt]: Never mind, didn't see the change to discriminant types.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -14212,12 +14222,26 @@ namespace ts {
             return undefined;
         }
 
+        function isDiscriminantType(type: Type): boolean {
```
Comment (sandersn) [inline src/compiler/checker.ts]: this seems unrelated to the bug fix in the PR description

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -14212,12 +14222,26 @@ namespace ts {
...
         }
 
+        function isDiscriminantType(type: Type): boolean {
+            if (type.flags & TypeFlags.Union) {
+                if (type.flags & (TypeFlags.Boolean | TypeFlags.EnumLiteral)) {
+                    return true;
+                }
+                let combined = 0;
+                for (const t of (<UnionType>type).types) combined |= t.flags;
+                if (combined & TypeFlags.Unit && !(combined & TypeFlags.Instantiable)) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: what happens if you have two two non-unit types? Should that still be discriminable?

Comment (weswigham) [reply src/compiler/checker.ts]: I'm pretty sure this change isn't supposed to be here, since it's https://github.com/Microsoft/TypeScript/pull/27695

**Comment by ahejlsberg** [comment]
I accidentally based this PR on the `mixedDiscriminantTypes` branch (which has been merged now). Just look at the commits starting with "Handle circular mapped type instantiations for arrays and tuples".
